File: CS/SnapModule/Controllers/DocumentDataController.cs
using DevExpress.ExpressApp;
using SnapModule.Base;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.SystemModule;
using SnapModule.Helpers;

namespace SnapModule.Controllers {
    public class DocumentDataController : ViewController {
        SimpleAction showDocumentDesignerAction, showDocumentPrintPreviewAction, editDocumentDataAction;
        NewObjectViewController newObjectViewController;
        ListViewProcessCurrentObjectController processCurrentObjectController;
        public DocumentDataController() {
            TargetObjectType = typeof(IDocumentData);
            TargetViewType = ViewType.ListView;

            showDocumentDesignerAction = new SimpleAction(this, "ShowDocumentDesigner", DevExpress.Persistent.Base.PredefinedCategory.View);
            showDocumentDesignerAction.Caption = "Show Designer";
            showDocumentDesignerAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject;
            showDocumentDesignerAction.ImageName = "Action_Report_ShowDesigner";
            showDocumentDesignerAction.Execute += new SimpleActionExecuteEventHandler(showDocumentDesignerAction_Execute);

            showDocumentPrintPreviewAction = new SimpleAction(this, "ShowDocumentPrintPreview", DevExpress.Persistent.Base.PredefinedCategory.View);
            showDocumentPrintPreviewAction.Caption = "Show Print Preview";
            showDocumentPrintPreviewAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject;
            showDocumentPrintPreviewAction.ImageName = "Action_Printing_Preview";
            showDocumentPrintPreviewAction.Execute += new SimpleActionExecuteEventHandler(showDocumentPrintPreviewAction_Execute);

            editDocumentDataAction = new SimpleAction(this, "EditDocumentData", DevExpress.Persistent.Base.PredefinedCategory.View);
            editDocumentDataAction.Caption = "Edit Document Data";
            editDocumentDataAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject;
            editDocumentDataAction.ImageName = "Action_Edit";
            editDocumentDataAction.Execute += new SimpleActionExecuteEventHandler(editDocumentDataAction_Execute);
        }
        protected override void OnActivated() {
            base.OnActivated();
            newObjectViewController = Frame.GetController<NewObjectViewController>();
            if (newObjectViewController != null) {
                newObjectViewController.ObjectCreating += DocumentDataController_ObjectCreating;
            }
            processCurrentObjectController = Frame.GetController<ListViewProcessCurrentObjectController>();
            if (processCurrentObjectController != null) {
                processCurrentObjectController.CustomProcessSelectedItem += processCurrentObjectController_CustomProcessSelectedItem;
            }
        }
        protected override void OnDeactivated() {
            base.OnDeactivated();
            if (newObjectViewController != null) {
                newObjectViewController.ObjectCreating -= DocumentDataController_ObjectCreating;
                newObjectViewController = null;
            }
            if (processCurrentObjectController != null) {
                processCurrentObjectController.CustomProcessSelectedItem -= processCurrentObjectController_CustomProcessSelectedItem;
                processCurrentObjectController = null;
            }
        }
        private void DocumentDataController_ObjectCreating(object sender, ObjectCreatingEventArgs e) {
            if (!e.Cancel && typeof(IDocumentData).IsAssignableFrom(e.ObjectType) &&
                Frame.Context != TemplateContext.LookupControl && Frame.Context != TemplateContext.LookupWindow) {
                e.ShowDetailView = false;
                IDocumentData documentData = (IDocumentData)e.ObjectSpace.CreateObject(e.ObjectType);
                e.NewObject = documentData;
                ShowDesigner(documentData, e.ObjectSpace);
            }
        }
        private void processCurrentObjectController_CustomProcessSelectedItem(object sender, CustomProcessListViewSelectedItemEventArgs e) {
            if (!e.Handled && Frame.Context != TemplateContext.LookupControl && Frame.Context != TemplateContext.LookupWindow &&
                showDocumentDesignerAction.Active && showDocumentDesignerAction.Enabled) {
                e.Handled = true;
                showDocumentDesignerAction.DoExecute();
            }
        }
        private void showDocumentDesignerAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            IObjectSpace newObjectSpace = Application.CreateObjectSpace(View.ObjectTypeInfo.Type);
            ShowDesigner((IDocumentData)newObjectSpace.GetObject(View.CurrentObject), newObjectSpace);
        }
        private void ShowDesigner(IDocumentData documentData, IObjectSpace objectSpace) {
            objectSpace.Committed += delegate { ObjectSpace.Refresh(); };
            SnapDocumentHelper.ShowDesigner(documentData, objectSpace, Application);
        }
        private void showDocumentPrintPreviewAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            IObjectSpace newObjectSpace = Application.CreateObjectSpace(View.ObjectTypeInfo.Type);
            SnapDocumentHelper.ShowPrintPreview((IDocumentData)newObjectSpace.GetObject(View.CurrentObject), newObjectSpace);
        }
        private void editDocumentDataAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            ListViewProcessCurrentObjectController.ShowObject(e.CurrentObject, e.ShowViewParameters, Application, Frame, View);
            e.ShowViewParameters.CreatedView.ObjectSpace.Committed += delegate { ObjectSpace.Refresh(); };
        }
    }
}

File: VB/SnapModule/Controllers/DocumentDataController.vb
Imports DevExpress.ExpressApp
Imports SnapModule.Base
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.SystemModule
Imports SnapModule.Helpers

Namespace SnapModule.Controllers
    Public Class DocumentDataController
        Inherits ViewController

        Private showDocumentDesignerAction, showDocumentPrintPreviewAction, editDocumentDataAction As SimpleAction
        Private newObjectViewController As NewObjectViewController
        Private processCurrentObjectController As ListViewProcessCurrentObjectController
        Public Sub New()
            TargetObjectType = GetType(IDocumentData)
            TargetViewType = ViewType.ListView

            showDocumentDesignerAction = New SimpleAction(Me, "ShowDocumentDesigner", DevExpress.Persistent.Base.PredefinedCategory.View)
            showDocumentDesignerAction.Caption = "Show Designer"
            showDocumentDesignerAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject
            showDocumentDesignerAction.ImageName = "Action_Report_ShowDesigner"
            AddHandler showDocumentDesignerAction.Execute, AddressOf showDocumentDesignerAction_Execute

            showDocumentPrintPreviewAction = New SimpleAction(Me, "ShowDocumentPrintPreview", DevExpress.Persistent.Base.PredefinedCategory.View)
            showDocumentPrintPreviewAction.Caption = "Show Print Preview"
            showDocumentPrintPreviewAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject
            showDocumentPrintPreviewAction.ImageName = "Action_Printing_Preview"
            AddHandler showDocumentPrintPreviewAction.Execute, AddressOf showDocumentPrintPreviewAction_Execute

            editDocumentDataAction = New SimpleAction(Me, "EditDocumentData", DevExpress.Persistent.Base.PredefinedCategory.View)
            editDocumentDataAction.Caption = "Edit Document Data"
            editDocumentDataAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject
            editDocumentDataAction.ImageName = "Action_Edit"
            AddHandler editDocumentDataAction.Execute, AddressOf editDocumentDataAction_Execute
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            newObjectViewController = Frame.GetController(Of NewObjectViewController)()
            If newObjectViewController IsNot Nothing Then
                AddHandler newObjectViewController.ObjectCreating, AddressOf DocumentDataController_ObjectCreating
            End If
            processCurrentObjectController = Frame.GetController(Of ListViewProcessCurrentObjectController)()
            If processCurrentObjectController IsNot Nothing Then
                AddHandler processCurrentObjectController.CustomProcessSelectedItem, AddressOf processCurrentObjectController_CustomProcessSelectedItem
            End If
        End Sub
        Protected Overrides Sub OnDeactivated()
            MyBase.OnDeactivated()
            If newObjectViewController IsNot Nothing Then
                RemoveHandler newObjectViewController.ObjectCreating, AddressOf DocumentDataController_ObjectCreating
                newObjectViewController = Nothing
            End If
            If processCurrentObjectController IsNot Nothing Then
                RemoveHandler processCurrentObjectController.CustomProcessSelectedItem, AddressOf processCurrentObjectController_CustomProcessSelectedItem
                processCurrentObjectController = Nothing
            End If
        End Sub
        Private Sub DocumentDataController_ObjectCreating(ByVal sender As Object, ByVal e As ObjectCreatingEventArgs)
            If (Not e.Cancel) AndAlso GetType(IDocumentData).IsAssignableFrom(e.ObjectType) AndAlso Frame.Context <> TemplateContext.LookupControl AndAlso Frame.Context <> TemplateContext.LookupWindow Then
                e.ShowDetailView = False
                Dim documentData As IDocumentData = DirectCast(e.ObjectSpace.CreateObject(e.ObjectType), IDocumentData)
                e.NewObject = documentData
                ShowDesigner(documentData, e.ObjectSpace)
            End If
        End Sub
        Private Sub processCurrentObjectController_CustomProcessSelectedItem(ByVal sender As Object, ByVal e As CustomProcessListViewSelectedItemEventArgs)
            If (Not e.Handled) AndAlso Frame.Context <> TemplateContext.LookupControl AndAlso Frame.Context <> TemplateContext.LookupWindow AndAlso showDocumentDesignerAction.Active.ResultValue AndAlso showDocumentDesignerAction.Enabled.ResultValue Then
                e.Handled = True
                showDocumentDesignerAction.DoExecute()
            End If
        End Sub
        Private Sub showDocumentDesignerAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim newObjectSpace As IObjectSpace = Application.CreateObjectSpace()
            ShowDesigner(DirectCast(newObjectSpace.GetObject(View.CurrentObject), IDocumentData), newObjectSpace)
        End Sub
        Private Sub ShowDesigner(ByVal documentData As IDocumentData, ByVal objectSpace As IObjectSpace)
            AddHandler objectSpace.Committed, Sub() ObjectSpace.Refresh()
            SnapDocumentHelper.ShowDesigner(documentData, objectSpace, Application)
        End Sub
        Private Sub showDocumentPrintPreviewAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim newObjectSpace As IObjectSpace = Application.CreateObjectSpace()
            SnapDocumentHelper.ShowPrintPreview(DirectCast(newObjectSpace.GetObject(View.CurrentObject), IDocumentData), newObjectSpace)
        End Sub
        Private Sub editDocumentDataAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            ListViewProcessCurrentObjectController.ShowObject(e.CurrentObject, e.ShowViewParameters, Application, Frame, View)
            AddHandler e.ShowViewParameters.CreatedView.ObjectSpace.Committed, Sub() ObjectSpace.Refresh()
        End Sub
    End Class
End Namespace
